### PR TITLE
comp: when targeting any Apple platform from macOS, always prefer native paths and libc installation

### LINKED ISSUE #16557: Clang errors when cross-compiling to aarch64-macos with C library dependencies
### Zig Version

0.11.0-dev.4059+17255bed4

### Steps to Reproduce and Observed Behavior

I uploaded a minimal example project that you can `git clone` to reproduce the issue:
https://github.com/konsuko/zraylibtest

Please note that I am using macOS with the M1 CPU, version 13.3.1, but I believe this issues also appears on Windows (when a friend tested the same source code)

### Expected Behavior

I am on macOS Ventura 13.3.1 (latest) and have a problem:

When compiling a minimal example project that uses [Raylib](https://github.com/raysan5/raylib) (a C library for game development), I am encountering an error that I think might be a compiler bug:

When using `zig build` or `zig build run`, everything works as expected and I have no errors: A window with a white background and an FPS counter opens up.

When I try to (cross) compile from/to aarch64 macOS by using `zig build -Dtarget=aarch64-macos`, I am getting clang errors:

```
zig build-lib raylib Debug aarch64-macos: error: error(compilation): clang failed with stderr: In file included from /Users/konsuko/dev/misc/zraylibtest/deps/raylib/src/rglfw.c:61:
In file included from /Users/konsuko/dev/misc/zraylibtest/deps/raylib/src/external/glfw/src/init.c:30:
In file included from /Users/konsuko/dev/misc/zraylibtest/deps/raylib/src/external/glfw/src/internal.h:331:
In file included from /Users/konsuko/dev/misc/zraylibtest/deps/raylib/src/external/glfw/src/platform.h:42:
/Users/konsuko/dev/misc/zraylibtest/deps/raylib/src/external/glfw/src/cocoa_platform.h:29:10: fatal error: 'Carbon/Carbon.h' file not found

error(compilation): clang failed with stderr: In file included from /Users/konsuko/dev/misc/zraylibtest/deps/raylib/src/raudio.c:176:
/Users/konsuko/dev/misc/zraylibtest/deps/raylib/src/external/miniaudio.h:31754:10: fatal error: 'CoreAudio/CoreAudio.h' file not found


zig build-lib raylib Debug aarch64-macos: error: the following command failed with 2 compilation errors:
/Users/konsuko/dev/zig/zig build-lib -cflags -std=gnu99 -D_GNU_SOURCE -DGL_SILENCE_DEPRECATION=199309L -fno-sanitize=undefined -- /Users/konsuko/dev/misc/zraylibtest/deps/raylib/src/raudio.c /Users/konsuko/dev/misc/zraylibtest/deps/raylib/src/rcore.c /Users/konsuko/dev/misc/zraylibtest/deps/raylib/src/rmodels.c /Users/konsuko/dev/misc/zraylibtest/deps/raylib/src/rshapes.c /Users/konsuko/dev/misc/zraylibtest/deps/raylib/src/rtext.c /Users/konsuko/dev/misc/zraylibtest/deps/raylib/src/rtextures.c /Users/konsuko/dev/misc/zraylibtest/deps/raylib/src/utils.c -cflags -std=gnu99 -D_GNU_SOURCE -DGL_SILENCE_DEPRECATION=199309L -fno-sanitize=undefined -ObjC -- /Users/konsuko/dev/misc/zraylibtest/deps/raylib/src/rglfw.c -lc --cache-dir /Users/konsuko/dev/misc/zraylibtest/zig-cache --global-cache-dir /Users/konsuko/.cache/zig --name raylib -static -target aarch64-macos -mcpu generic -I /Users/konsuko/dev/misc/zraylibtest/deps/raylib/src/external/glfw/include -D PLATFORM_DESKTOP -framework CoreServices -framework AppKit -framework CoreGraphics -framework Foundation -framework IOKit --listen=- 
Build Summary: 1/5 steps succeeded; 1 failed (disable with --summary none)
install transitive failure
└─ install zparser transitive failure
   └─ zig build-exe zparser Debug aarch64-macos transitive failure
      └─ zig build-lib raylib Debug aarch64-macos 2 errors
/Users/konsuko/dev/misc/zraylibtest/deps/raylib/src/rglfw.c:1:1: error: unable to build C object: clang exited with code 1
/Users/konsuko/dev/misc/zraylibtest/deps/raylib/src/raudio.c:1:1: error: unable to build C object: clang exited with code 1
```

I did some digging through other issues on Github and found this similar issue: https://github.com/ziglang/zig/issues/16235
I am not sure, but it looks like maybe the compiler is not finding macOS dylibs, even though they should exist on my machine (I have XCode command line tools installed and updated to the latest version).

Probably something that @kubkon would know more about?

I apologize if it's just me doing something dumb here and possibly an easy fix, I am very new to Zig and still learning :)

### PR DESCRIPTION
**Pull request by kubkon**
This change makes targeting iOS/watchOS/tvOS and macOS as a cross-compilation target from a macOS host use the system SDK by default before falling back to Zig shipped libc headers (if available).

Closes #16557

### CODE REVIEW

**Comment by kubkon** [comment]
Hmm, it is unfortunate that with this change it is impossible to test LLVM targets targeting `ios` (see CI failures). On the other hand, it is my understanding that targeting Apple platforms that are not macOS always requires the SDK, which would make the change proposed in this PR a desirable change. @mitchellh, @floooh and @slimsag I know the three of you will have some experience and opinion on this I would love to learn: do you think we should autodetect iOS/tvOS/watchOS SDKs automatically when compiling on macOS (like it's done with Xcode), or would you say we should treat it like any other cross-compilation target requiring the user to manually specify the search paths to the relevant SDK?

**Comment by kubkon** [comment]
On the other hand, perhaps current behavior of treating other Apple platforms as foreign ain't that bad. If the user wishes to provide iOS SDK (or similar) they can do this via `std.zig.system.darwin.getSdk(..., ios_target)` or via a build flag or similar. After all, the change in this PR assumes the user will have Xcode installed as this is the official way of obtaining SDKs for every Apple platform and as we all well know, Xcode is simply horrible software and if I didn't have to install it but rather use slim-Zig as an alternative, I would be absolutely delighted.

I think I am leaning towards scrapping this change, but I'll let y'all and anyone else chime in first before closing the PR. Perhaps I should have opened a proposal/question issue where we would discuss rather than opening a PR, but I was trying to fix another yet related issue.

**Comment by floooh** [comment]
I have been summoned? :)

I don't yet "officially" support iOS in my Zig stuff, but I was under the impression that cross-compiling to iOS requires to provide a sysroot to the iOS SDK. Currently I do this via xcrun when calling zig build, e.g.:

```
zig build --sysroot $(xcrun --sdk iphoneos --show-sdk-path) -Dtarget=aarch64-ios
```

...or for the simulator:

```
zig build --sysroot $(xcrun --sdk iphonesimulator --show-sdk-path) -Dtarget=aarch64-ios-simulator
```

Currently when I do that in https://github.com/floooh/pacman.zig with Zig 0.11.0 I get an error though that the C code can't find iOS framework headers, like:

```
.../sokol_app.h:1942:17: fatal error: 'Metal/Metal.h' file not found
```

I didn't get around to investigate yet, if you want to try yourself here are the instructions:

https://github.com/floooh/pacman.zig#experimental-ios-support

TL;DR: IMHO it's fine that a sysroot to a separate iOS SDK must be provided, but then it should also work for building C code (but that's probably more of a build system issue).

Btw: AFAIK a working Apple compiler toolchain can also be installed without Xcode, I think that's what homebrew is doing. Not entirely sure if this includes macOS/iOS SDKs though, or only C and C++ headers and runtime libs.

PPS: ideally it would also be nice to have an "iOS SDK" Zig package in the future, which includes all headers, libraries and frameworks to build iOS applications without the official SDK. This would be a killer feature for cross-compiling on non-Apple platforms.

**Comment by kubkon** [comment]
> I have been summoned? :)
> 
> I don't yet "officially" support iOS in my Zig stuff, but I was under the impression that cross-compiling to iOS requires to provide a sysroot to the iOS SDK anyway. Currently I do this via xcrun when calling zig build (I wasn't aware that the zig stdlib also has a helper function for finding an SDK)
> 
> ```
> zig build --sysroot $(xcrun --sdk iphoneos --show-sdk-path) -Dtarget=aarch64-ios
> ```
> 
> ...or for the simulator:
> 
> ```
> zig build --sysroot $(xcrun --sdk iphonesimulator --show-sdk-path) -Dtarget=aarch64-ios-simulator
> ```
> 
> Currently when I do that in https://github.com/floooh/pacman.zig with Zig 0.11.0 I get an error though that the ObjC code can't find iOS framework headers, like:
> 
> ```
> .../sokol_app.h:1942:17: fatal error: 'Metal/Metal.h' file not found
> ```
> 
> ...at some point in the past this had worked.

I am actually aware of this and if we land https://github.com/ziglang/zig/pull/16818, it will then require you to provide fully composed search paths but will get rid of the error. I spent a significant amount of time trying to understand why `-iwithsysroot` and `-iframeworkwithsysroot` stopped working but clearly have, and to make it more interesting, using system `clang` with explicitly passed sysroot, those flags and `-nostdinc` fails in exactly the same way.

**Comment by kubkon** [comment]
I will close this PR but feel free to drop a comment here.
